onnxruntime/core/providers/cuda/shared_inc/cuda_utils.h

```diff
@@ -33,5 +34,25 @@ std::unique_ptr<IConstantBuffer<T>> CreateConstantOnes();
 template <typename T>
 void Fill(T* output, T value, int64_t count);
 
+/*
+  This is a utility wrapper for arbitrary type array
+  Commonly used for passing small list of metadata during cuda kernel launch
+  It's better to pass the array by value than having another cuMemcpy to pass the data to device.
+*/
+template <typename T, int32_t capacity = 8>
+struct TArray {
+  TArray() : size_(0), data_() {
```

Comment:
Scott McKay
Does 'size' need to be <= MAX_ARRAY_SIZE? If so there should probably be some check of that.

Sherlock Huang
yes. ORT_ENFORCE